### PR TITLE
RtpsRelay: Implement rapid association

### PR DESCRIPTION
**Pull request by jrw972**

### CODE REVIEW

File: dds/DCPS/RTPS/Spdp.cpp [hunk 1]
```diff
@@ -2317,6 +2316,28 @@ Spdp::Interceptor::reactor_is_shut_down() const {
...
+    ScheduleCommand(Spdp* spdp, const ACE_Time_Value& delay) : spdp_(spdp), delay_(delay) {}
+
+  private:
+    void execute() {
+      if (-1 == spdp_->tport_->reactor()->schedule_timer(spdp_->tport_, this, delay_)) {
+        throw std::runtime_error("failed to schedule timer with reactor");
+      }
+    }
+
+    Spdp* const spdp_;
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp]: `const Spdp* spdp_`?

File: dds/DCPS/RTPS/Spdp.cpp [hunk 2]
```diff
@@ -2317,6 +2316,28 @@ Spdp::Interceptor::reactor_is_shut_down() const {
   return task_->is_shut_down();
 }
 
+void
+Spdp::schedule_send(const ACE_Time_Value& delay)
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp]: Could you use `TimeDuration` for this? I would like use to not use `ACE_Time_Value` directly in the core libraries unless interacting with ACE.

File: dds/DCPS/RTPS/Spdp.cpp (hunk 1 above)
Comment (mitza-oci) [reply dds/DCPS/RTPS/Spdp.cpp]: Moving `const` to the other side of `*` changes its meaning

Comment (iguessthislldo) [reply dds/DCPS/RTPS/Spdp.cpp]: I forgot how to read pointers again. I was thinking like it was `Spdp const*` or `const Spdp*`. The next line should definitely be changed though.

Comment (simpsont-oci) [reply dds/DCPS/RTPS/Spdp.cpp]: This is the baitclick of github review comments... I can't stand the suspense. Now I have to open the file to figure out what the next line is!

Comment (oschwaldp-oci) [reply dds/DCPS/RTPS/Spdp.cpp]: clicked...and opened file to see next line.

File: dds/DCPS/RTPS/Spdp.cpp [hunk 3]
```diff
@@ -2317,6 +2316,28 @@ Spdp::Interceptor::reactor_is_shut_down() const {
...
+
+  private:
+    void execute() {
+      if (-1 == spdp_->tport_->reactor()->schedule_timer(spdp_->tport_, this, delay_)) {
+        throw std::runtime_error("failed to schedule timer with reactor");
+      }
+    }
+
+    Spdp* const spdp_;
+    ACE_Time_Value const delay_;
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp]:
```suggestion
    const ACE_Time_Value delay_;
```

File: tools/rtpsrelay/AssociationTable.cpp [hunk 4]
```diff
@@ -4,16 +4,19 @@
 
 namespace RtpsRelay {
 
-void AssociationTable::insert(const WriterEntry& writer_entry)
+void AssociationTable::insert(const WriterEntry& writer_entry,
+                              GuidSet& local_guids,
+                              RelayAddressesSet& relay_addresses)
 {
+  const bool local = writer_entry.relay_addresses() == relay_addresses_;
```
Comment (mitza-oci) [inline tools/rtpsrelay/AssociationTable.cpp]:
```suggestion
  const auto local = writer_entry.relay_addresses() == relay_addresses_;
```

File: tools/rtpsrelay/PublicationListener.cpp [hunk 5]
```diff
@@ -6,10 +6,10 @@ namespace RtpsRelay {
 
 PublicationListener::PublicationListener(OpenDDS::DCPS::DomainParticipantImpl* participant,
                                          WriterEntryDataWriter_ptr writer,
-                                         const AssociationTable& association_table) :
+                                         const RelayAddresses& relay_addresses) :
```
Comment (mitza-oci) [inline tools/rtpsrelay/PublicationListener.cpp]:
```suggestion
                                         const RelayAddresses& relay_addresses)
```
(see ReaderListener::ReaderListener)

File: tools/rtpsrelay/RelayHandler.h [hunk 6]
```diff
@@ -71,9 +70,10 @@ class VerticalHandler : public RelayHandler {
 
 protected:
   virtual std::string extract_relay_address(const RelayAddresses& relay_addresses) const = 0;
-  virtual void add_addresses(const ACE_INET_Addr& /*a_remote*/,
-                             const OpenDDS::DCPS::RepoId& /*a_src_guid*/,
-                             std::set<std::string>& /* a_addrs*/) {}
+  virtual bool do_normal_processing (const ACE_INET_Addr& /*a_remote*/,
```
Comment (mitza-oci) [inline tools/rtpsrelay/RelayHandler.h]:
```suggestion
  virtual bool do_normal_processing(const ACE_INET_Addr& /*a_remote*/,
```

File: tools/rtpsrelay/RelayHandler.h [hunk 7]
```diff
@@ -71,9 +70,10 @@ class VerticalHandler : public RelayHandler {
...
+                                     const OpenDDS::DCPS::RepoId& /*a_src_guid*/,
+                                     ACE_Message_Block* /*a_msg*/) { return true;}
```
Comment (mitza-oci) [inline tools/rtpsrelay/RelayHandler.h]:
```suggestion
                                     ACE_Message_Block* /*a_msg*/) { return true; }
```

File: tools/rtpsrelay/RtpsRelay.cpp [hunk 8]
```diff
@@ -247,6 +243,45 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
 
   AssociationTable association_table(relay_addresses);
 
+  HorizontalHandler spdp_horizontal_handler(reactor, association_table);
+  HorizontalHandler sedp_horizontal_handler(reactor, association_table);
+  HorizontalHandler data_horizontal_handler(reactor, association_table);
+
+  OpenDDS::DCPS::RepoId application_participant_id = application_participant_impl->get_id();
```
Comment (mitza-oci) [inline tools/rtpsrelay/RtpsRelay.cpp]:
```suggestion
  const auto application_participant_id = application_participant_impl->get_id();
```

File: tools/rtpsrelay/RtpsRelay.cpp [hunk 9]
```diff
@@ -389,63 +438,46 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
...
-    << "SEDP Vertical listening on " << sedp_vertical_handler.relay_address() << '\n'
-    << "Data Vertical listening on " << data_vertical_handler.relay_address() << std::endl;
-
   StatisticsHandler statistics_h(reactor,
                                  &spdp_vertical_handler, &spdp_horizontal_handler,
                                  &sedp_vertical_handler, &sedp_horizontal_handler,
                                  &data_vertical_handler, &data_horizontal_handler);
   statistics_h.open();
 
+  rtps_discovery->schedule_send(application_domain, application_participant_id, ACE_Time_Value(1,0));
```
Comment (mitza-oci) [inline tools/rtpsrelay/RtpsRelay.cpp]:
```suggestion
  rtps_discovery->schedule_send(application_domain, application_participant_id, ACE_Time_Value(1));
```

File: tools/rtpsrelay/WriterListener.h [hunk 10]
```diff
@@ -3,18 +3,20 @@
...
 #include "ListenerBase.h"
-#include "lib/RelayTypeSupportImpl.h"
+#include "RelayHandler.h"
 
 namespace RtpsRelay {
 
 class WriterListener : public ListenerBase {
 public:
-  explicit WriterListener(AssociationTable& association_table);
+  explicit WriterListener(AssociationTable& association_table,
```
Comment (mitza-oci) [inline tools/rtpsrelay/WriterListener.h]:
```suggestion
  WriterListener(AssociationTable& association_table,
```

File: tools/rtpsrelay/lib/QosIndex.h [hunk 11]
```diff
@@ -184,20 +310,28 @@ class Literal {
...
+    index_.insert(reader, local_guids, relay_addresses);
   }
 
-  void insert_pattern(ReaderPtr reader)
+  void insert_pattern(ReaderPtr reader,
+                      GuidSet& local_guids,
+                      RelayAddressesSet& relay_addresses)
   {
-    index_.insert(reader);
+    return index_.insert(reader, local_guids, relay_addresses);
```
Comment (mitza-oci) [inline tools/rtpsrelay/lib/QosIndex.h]:
```suggestion
    index_.insert(reader, local_guids, relay_addresses);
```

Comment (jrw972) [reply tools/rtpsrelay/lib/QosIndex.h]: You are better than a compiler.

File: dds/DCPS/RTPS/Spdp.cpp (hunk 2 above)
Comment (jrw972) [reply dds/DCPS/RTPS/Spdp.cpp]: Absolutely.

File: tools/rtpsrelay/lib/QosIndex.h (hunk 11 above)
Comment (mitza-oci) [reply tools/rtpsrelay/lib/QosIndex.h]: In some ways yes, in others, no.  My debuginfo is OK but inter-procedural optimization not so much.

**Pull request merged**